projs/shadow/shadow-engine/core/inc/event-bus/event_bus.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <string>
+#include <functional>
+#include <map>
+#include <memory>
+#include <any>
+
+#include "SHObject.h"
+#include "events.h"
+
+namespace ShadowEngine::EventBus {
+
+    template<class T, class U>
+    concept Derived = std::is_base_of<T, U>::value;
+
+    template<typename T>
+    concept SHObjectConcept = Derived<ShadowEngine::SHObject, T>;
+
+
+
+
+    template<typename T>
+    concept EventType = Derived<Event, T>;
+
+    template<EventType T>
+    struct Subscription {
+        using Callback = std::function<const void(T &)>;
+
+        Callback callback;
+
+        std::weak_ptr<ShadowEngine::SHObject> binding;
+
+        explicit Subscription(Callback call): callback(call){}
+    };
+
+    class EventDispatcherBase {
+    public:
+        virtual ~EventDispatcherBase() = default;
+
+        virtual void call(ShadowEngine::SHObject &obj) = 0;
+    };
+
+    template<EventType T>
+    class EventDispatcher : public EventDispatcherBase {
+    public:
+        using SubRef = std::shared_ptr<Subscription<T>>;
+
+    protected:
+        using SubscriptionList = std::vector<SubRef>;
```

Comment:
Vector was used here because I'm not sure how many subscribers there will be. As far as I understand vector reallocates with increasingly larger free space. I don't think this would impact us too much here. Maybe a pre-sizing to some large number could solve it. If not then not sure what other container I could use. (Suggestions welcome)

Reply:
Vector reallocates with increasing space, yes, but the initial assignment (up to ~50 elements) causes rapid memory churn as vector keeps destroying and allocating memory in a tight loop. 

That'll kill performance for us.